### PR TITLE
Update WP compatibility check in `gutenberg_pre_init()`

### PR DESCRIPTION
**Pull request by jeremyfelt**
## Description

This updates the `version_compare()` check in `gutenberg_pre_init()` in a few ways:

1. Move from requiring 5.3 to 5.6. The plugin relies on `WP_Block_Supports`, which was added in WP 5.6, and `register_block_format()`, which was added in WP 5.5.
2. Fix the version comparison. WordPress reports its version as `5.7` rather than `5.7.0`. `version_compare( '5.7', '5.7.0', '<' )` evaluates as `true`, which puts 5.7.1 as the actual minimum version. When compatibility is checked against a major WP release, the X.Y version number should be used. See previously, #15565.
3. After adding an admin notice, the plugin bails with `return;` and Gutenberg libraries are not loaded. Other hooks are added by the plugin that require those libraries, and a fatal error occurs rather than a compatibility notice. This removes the `return`, which helps to eliminate immediate fatal errors and allows the notice to be shown.

A (likely preferable) alternative to item 3 would be to remove all other functionality from the main plugin file so that it too is only loaded when `gutenberg_pre_init()` fires successfully.

Current culprits appear to be:
* `modify_welcome_panel()`
* `modify_admin_bar()`
* `gutenberg_site_editor_menu()`
* `gutenberg_menu()`

## How has this been tested?

I've tested by updating core back and forth between WP 5.3, 5.5, 5.6, and 5.7 while the plugin is in an active an inactive state.

## Types of changes

Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [NA] My code follows the accessibility standards. 
- [NA] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [NA] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by jeremyfelt** [comment]
I don't see how the end-to-end test failure is related to this change and I don't have the capability to restart the job to see if it's just a one-off issue. Happy to fix anything if needed! :)

**Comment by gziolo** [comment]
> I don't see how the end-to-end test failure is related to this change and I don't have the capability to restart the job to see if it's just a one-off issue. Happy to fix anything if needed! :)

Yes, it looks unrelated. E2E tests don't run with the older versions of WordPress. I re-triggered those tests.

**Comment by paaljoachim** [comment]
Hey

I just want to say that I tried activating the Gutenberg plugin 10.2.1 on an older WP site that is running 5.3.6. 
It would not activate and gave the message that the plugin could not be activated as a fatal error was detected (loosely translated from Norwegian). 
It would be helpful if it also mentioned update to the newest version of WordPress to be able to use a new version of Gutenberg.

https://user-images.githubusercontent.com/5323259/111844489-6bb71580-8903-11eb-8dda-1cb46084644a.mp4

File: gutenberg.php
Comment (gziolo) [inline gutenberg.php:209]: With the `return` removed it will try to load all the files for the plugin even for older versions that aren't supported. Won't it cause an issue?

Comment (gziolo) [reply gutenberg.php:209]: I see an answer:

> After adding an admin notice, the plugin bails with return; and Gutenberg libraries are not loaded. Other hooks are added by the plugin that require those libraries, and a fatal error occurs rather than a compatibility notice. This removes the return, which helps to eliminate immediate fatal errors and allows the notice to be shown.

Should we move all hooks to their own file that gets loaded first in `load.php`?

**Comment by gziolo** [comment]
I see another line where `5.3` is hardcoded:
https://github.com/WordPress/gutenberg/blob/b8bb4fbef4a268cbbb1f46a0b14d953e5b0da173/gutenberg.php#L172

We should probably move the version to the constant to make it easier to upgrade in the future.

File: gutenberg.php
Comment (gziolo) [reply gutenberg.php:209]: https://github.com/WordPress/gutenberg/pull/30230 should resolve the issue with errors so we can keep `return;` here.

**Comment by swissspidy** [comment]
Previously reported in #27998

**Comment by jeremyfelt** [comment]
I'll update the PR to add the `return` statement back in. #30230 is excellent. 👍🏻 

> We should probably move the version to the constant to make it easier to upgrade in the future.

https://github.com/WordPress/gutenberg/blob/da8555d258edbe676fa079fb51252f918ae032e1/bin/generate-gutenberg-php.php#L19

Does it complicate things too much to add this to `print_production_defines()`? If I understand how that file is used, it could then be automatically generated based on the plugin header's required version.

> Previously reported in #27998

I completely missed this! 🤦🏻

**Comment by gziolo** [comment]
I landed https://github.com/WordPress/gutenberg/pull/30230 to unblock this PR.

> Does it complicate things too much to add this to `print_production_defines()`? If I understand how that file is used, it could then be automatically generated based on the plugin header's required version.

I thought about a simpler approach where we define the constant in the file, but given that we need it only on production, we can take a more complex path 😄 I leave it up to you to decide. The solution you mentioned has the benefit of updating only the minimum version in the plugin's metadata moving forward. The downside is that we would not have access to this constant in development mode, so we would have to account for that somehow.

**Review by gziolo — APPROVED** [review]
This should be good to go now with the changes I applied.

**Comment by gziolo** [comment]
It works great now in my testing when using the latest `trunk`.

**Pull request merged**
